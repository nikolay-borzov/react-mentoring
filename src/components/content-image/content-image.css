.content-image {
  max-width: 100%;
  display: block;
}

.content-image-placeholder {
  position: relative;
  max-width: 100%;
  display: block;
  width: 100%;
  /* Keep aspect ratio */
  padding-top: 150%;
  background-color: rgba(0, 0, 0, 0.1);

  & > * {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }
}

.imageloader-failed {
  /* Take all available width (needed for keeping aspect ratio)*/
  width: 100%;
}