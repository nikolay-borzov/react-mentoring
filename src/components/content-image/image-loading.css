@import '../../styles/variables.css';


/* https://github.com/tobiasahlin/SpinKit */
.image-spinner {
  width: 5rem;
  height: 5rem;
}

.image-spinner__bounce1,
.image-spinner__bounce2 {
  width: 100%;
  height: 100%;
  border-radius: 50%;
  background-color: var(--color-default);
  opacity: 0.6;
  position: absolute;
  top: 0;
  left: 0;

  animation: sk-bounce 2s infinite ease-in-out;
}

.image-spinner__bounce2 {
  animation-delay: -1s;
}

@keyframes sk-bounce {
  0%,
  100% {
    transform: scale(0);
  }
  50% {
    transform: scale(1);
  }
}
