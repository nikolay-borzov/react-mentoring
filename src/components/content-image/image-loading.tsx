import React from 'react'
import styled, { keyframes } from 'styled-components'

/* Animation from https://github.com/tobiasahlin/SpinKit */
const ImageSpinner = styled.div`
  width: 5rem;
  height: 5rem;
`

const bounce = keyframes`
 0%,
  100% {
    transform: scale(0);
  }
  50% {
    transform: scale(1);
  }
`

const ImageSpinnerBounce1 = styled.div`
  width: 100%;
  height: 100%;
  border-radius: 50%;
  background-color: var(--color-default);
  opacity: 0.6;
  position: absolute;
  top: 0;
  left: 0;

  animation: ${bounce} 2s infinite ease-in-out;
`
const ImageSpinnerBounce2 = styled(ImageSpinnerBounce1)`
  animation-delay: -1s;
`

export const ImageLoading: React.FunctionComponent = () => (
  <ImageSpinner>
    <ImageSpinnerBounce1 />
    <ImageSpinnerBounce2 />
  </ImageSpinner>
)
