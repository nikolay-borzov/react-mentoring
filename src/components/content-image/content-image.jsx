import React from 'react'
import PropTypes from 'prop-types'

import './content-image.css'

import ImageLoader from 'react-load-image'
import { ImageLoading } from './image-loading'

ContentImage.propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
  title: PropTypes.string
}

ContentImage.defaultProps = {
  title: ''
}

export function ContentImage({ src, title, alt }) {
  return (
    <ImageLoader src={src}>
      <img className="content-image" title={title} alt={alt} />
      <div className="content-image-placeholder color-alt">
        <span>Image unavailable</span>
      </div>
      <ImageLoading className="content-image-placeholder" />
    </ImageLoader>
  )
}
