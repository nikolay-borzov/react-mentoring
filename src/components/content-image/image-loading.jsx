import React from 'react'
import PropTypes from 'prop-types'

import './image-loading.css'

ImageLoading.propTypes = {
  className: PropTypes.string
}

ImageLoading.defaultProps = {
  className: ''
}

export function ImageLoading(props) {
  return (
    <div className={`image-loading-indicator ${props.className}`}>
      <div className="image-spinner">
        <div className="image-spinner__bounce1" />
        <div className="image-spinner__bounce2" />
      </div>
    </div>
  )
}
