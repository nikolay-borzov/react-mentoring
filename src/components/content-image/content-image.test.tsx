import React from 'react'

import {
  itRendersCorrectlyShallow,
  itContainsComponent
} from '../../../jest/test-helpers'

import { ContentImage, ContentImageProps } from './content-image'

describe('ContentImage component', () => {
  let props: ContentImageProps

  beforeEach(() => {
    props = {
      src: 'http://images/test.png',
      alt: 'Image alt text',
      title: undefined
    }
  })

  itRendersCorrectlyShallow(() => <ContentImage {...props} />)

  itRendersCorrectlyShallow(() => {
    props.title = 'Image title'
    return <ContentImage {...props} />
  }, 'renders correctly with title specified')

  itContainsComponent(() => <ContentImage {...props} />, 'ImageLoading')
})
