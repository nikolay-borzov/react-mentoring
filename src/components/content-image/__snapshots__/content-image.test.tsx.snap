// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentImage component renders correctly 1`] = `
<ImageLoaderStyled
  src="http://images/test.png"
>
  <ContentImageStyled
    alt="Image alt text"
    title=""
  />
  <ContentImagePlaceholder
    className="color-alt"
  >
    <span>
      Image unavailable
    </span>
  </ContentImagePlaceholder>
  <ContentImagePlaceholder>
    <ImageLoading />
  </ContentImagePlaceholder>
</ImageLoaderStyled>
`;

exports[`ContentImage component renders correctly with title specified 1`] = `
<ImageLoaderStyled
  src="http://images/test.png"
>
  <ContentImageStyled
    alt="Image alt text"
    title="Image title"
  />
  <ContentImagePlaceholder
    className="color-alt"
  >
    <span>
      Image unavailable
    </span>
  </ContentImagePlaceholder>
  <ContentImagePlaceholder>
    <ImageLoading />
  </ContentImagePlaceholder>
</ImageLoaderStyled>
`;
